File: 2023-PFSD-Session-01/2023-PFSD-Session-01/src/main/java/eci/edu/co/Transaction.java
package eci.edu.co;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Transaction {

    private Double value;
    private String currency;
    private String type;

    public static final String COP = "COP";
    public static final String USD = "USD";
    public static final String EUR = "EUR";

    public static List<String> currencies = new ArrayList<>(List.of(COP, USD, EUR));
    public static List<String> types = new ArrayList<>(List.of("Income", "Expense"));

    public Transaction(Double value, String currency, String type) {
        this.value = value;
        this.currency = currency;
        this.type = type;
    }

    public Double getValueInCop() {
        switch (currency) {
            case USD:
                return value * 4098;
            case EUR:
                return value * 4455;
            case COP:
                return value;
            default:
                return 0.0;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Transaction that = (Transaction) o;
        return Objects.equals(value, that.value) && Objects.equals(currency, that.currency) && Objects.equals(type, that.type);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, currency, type);
    }

    @Override
    public String toString() {
        return "Value: " + value + ", Currency: " + currency + ", Type: " + type;
    }
}

File: CASO1/src/main/scala/caso1/entities/Transaction.scala
package caso1.entities

import Transaction._

import java.sql.Timestamp

case class Transaction(value: Double, currency: String, `type`: String = "", category: String, timestamp: Timestamp) {

  def this (currency: String, `type`: String, category: String, timestamp: Timestamp) = {
    this(0, currency, `type`, category, timestamp)
  }

  var listTransactions = List(this)

  //def valueInCop = calculateValueInCop(currency, value)

  //def valueInCopT = calculateValueInCop(this)

  def transactionsOfType = getTransactionsOfType(listTransactions, `type`)
  def transactionsOfCurrency = getTransactionsOfCurrency(listTransactions, currency)
  def sumTransOfType  = sumTransactionsOfType(listTransactions, `type`)
  def printTransaction: Unit = println(toString)

  override def toString: String = s"Value: $value, Currency: $currency, Type: ${`type`}"

}

object Transaction {

  import scala.util.Random

  val COP = "COP"
  val CLP = "CLP"
  val USD = "USD"
  val EUR = "EUR"
  val GBP = "GBP"
  val JPY = "JPY"
  val Currencies = List(COP, USD, EUR, GBP, JPY)

  val USD_COP = 4098
  val EUR_COP = 4455
  val GBP_COP = 5210
  val JPY_COP = 28

  val Income = "Income"
  val Expense = "Expense"
  val Types = List(Income, Expense)

  val random = new Random()

  val Bills = "Bills"
  val Transportation = "Transportation"
  val Shopping = "Shopping"
  val Entertainment = "Entertainment"
  val Food = "Food"
  val ExpenseCategories = List(Bills,Transportation, Shopping, Entertainment, Food)

  val Salary = "Salary"
  val Selling = "Selling"
  val Gifts = "Gifts"
  val Award = "Award"
  val IncomeCategories = List(Salary, Selling, Gifts, Award)

  val Categories = List(IncomeCategories, ExpenseCategories)

  def getValue(transaction: Transaction): Double = {
    transaction.value
  }

  def getCurrency(transaction: Transaction): String = {
    transaction.currency
  }

  def getType(transaction: Transaction): String = {
    transaction.`type`
  }

  def getCategory(transaction: Transaction): String = {
    transaction.category
  }

  def getTimestamp(transaction: Transaction): Timestamp = {
    transaction.timestamp
  }

  def getTransactionsOfType(transactions: List[Transaction], `type`: String): List[Transaction] = {
    val typeTransactions = List[Transaction]()
    for (t <- transactions) {
      if (t.`type` == `type`) return t :: typeTransactions
    }
    typeTransactions
  }

  def getTransactionsOfCurrency(transactions: List[Transaction], currency: String): List[Transaction] = {
    val currencyTransactions = Nil
    for (t <- transactions) {
      if (t.currency == currency) return t +: currencyTransactions
    }
    currencyTransactions
  }

  def sumTransactionsOfType(transactions: List[Transaction], `type`: String): Double = {
    var sum = 0
    for (t <- transactions) {
      if (t.`type` == `type`) sum += t.value
    }
    sum
  }
}
